In [1]:
import numpy as np
import torch
import torch.nn as nn
import matplotlib.pyplot as plt

# 设置matplotlib为黑色主题（与课件保持一致）
plt.style.use('dark_background')
plt.rcParams.update({
    'figure.facecolor': 'black',
    'axes.facecolor': 'black',
    'axes.edgecolor': 'white',
    'axes.labelcolor': 'white',
    'text.color': 'white',
    'xtick.color': 'white',
    'ytick.color': 'white',
    'grid.color': '#444444',
    'grid.alpha': 0.3,
    'legend.facecolor': 'black',
    'legend.edgecolor': 'white',
})

print('✅ 环境准备完成！')

✅ 环境准备完成！


# <font color="#00FFFF" > 多层感知机 - 课堂练习答案</font>

本文件包含课堂练习的参考答案和详细解析。

---

## <font color="#FFEA00" >练习1：概率与Logit的相互转换 - 答案</font>

### ✅ 答案表

| 序号 | 概率 p | Logit z = log(p/(1-p)) | Sigmoid验证 σ(z) |
|:---:|:---:|:---:|:---:|
| 1 | 0.5 | **0** | **0.5** |
| 2 | 0.8 | **1.386** | **0.8** |
| 3 | 0.2 | **-1.386** | **0.2** |
| 4 | **0.5** | 0 | **0.5** |
| 5 | **0.881** | 2 | **0.881** |
| 6 | **0.269** | -1 | **0.269** |

In [2]:
# ==========================================
# 练习1：答案验证代码
# ==========================================

def logit(p):
    """从概率计算logit"""
    return np.log(p / (1 - p))

def sigmoid(z):
    """从logit计算概率"""
    return 1 / (1 + np.exp(-z))

print('=== 练习1：答案验证 ===')
print()

# 从概率到logit
print('1. 概率 -> Logit:')
p_values = [0.5, 0.8, 0.2]
for p in p_values:
    z = logit(p)
    print(f'   p={p} -> z={z:.3f}')

print()

# 从logit到概率
print('2. Logit -> 概率:')
z_values = [0, 2, -1]
for z in z_values:
    p = sigmoid(z)
    print(f'   z={z} -> p={p:.3f}')

print()
print('✅ 验证完成！')

=== 练习1：答案验证 ===

1. 概率 -> Logit:
   p=0.5 -> z=0.000
   p=0.8 -> z=1.386
   p=0.2 -> z=-1.386

2. Logit -> 概率:
   z=0 -> p=0.500
   z=2 -> p=0.881
   z=-1 -> p=0.269

✅ 验证完成！


### 📖 详细解析

**计算过程说明：**

1. **p = 0.5 → z = 0**
   - z = log(0.5 / 0.5) = log(1) = **0**
   - σ(0) = 1 / (1 + e^0) = 1/2 = **0.5**

2. **p = 0.8 → z ≈ 1.386**
   - z = log(0.8 / 0.2) = log(4) ≈ **1.386**
   - σ(1.386) ≈ **0.8**

3. **p = 0.2 → z ≈ -1.386**
   - z = log(0.2 / 0.8) = log(0.25) ≈ **-1.386**
   - σ(-1.386) ≈ **0.2**

4. **z = 0 → p = 0.5**
   - p = σ(0) = 1 / (1 + 1) = **0.5**

5. **z = 2 → p ≈ 0.881**
   - p = σ(2) = 1 / (1 + e^(-2)) ≈ **0.881**

6. **z = -1 → p ≈ 0.269**
   - p = σ(-1) = 1 / (1 + e^1) ≈ **0.269**

---

## <font color="#FFEA00" >练习2：Softmax手动计算 - 答案</font>

### ✅ 任务1：基础Softmax计算

**步骤1：计算指数值**

| 类别 | Logit z_i | $e^{z_i}$ |
|:---:|:---:|:---:|
| 0 | 2.0 | **7.39** |
| 1 | 1.0 | **2.72** |
| 2 | 0.1 | **1.11** |
| | **求和** | **11.22** |

**步骤2：计算Softmax概率**

| 类别 | 计算过程 | Softmax概率 |
|:---:|:---|:---:|
| 0 | 7.389 / 11.213 | **0.659** |
| 1 | 2.718 / 11.213 | **0.242** |
| 2 | 1.105 / 11.213 | **0.099** |
| | **概率之和** | **1.000** |

**步骤3答案：**

**类别0**被预测为正确类别，因为其logit值最大（2.0），经过Softmax后概率最高（约65.9%）。

### ✅ 任务2：观察温度效应

**放大10倍（z' = [20, 10, 1]）：**

| 类别 | 原始Logit | 放大10倍 z' | $e^{z'}$ | Softmax概率 |
|:---:|:---:|:---:|:---:|:---:|
| 0 | 2.0 | 20 | **4.85×10⁸** | **≈1.000** |
| 1 | 1.0 | 10 | **22026.5** | **≈0.000** |
| 2 | 0.1 | 1 | **2.718** | **≈0.000** |

**步骤2答案：**

当logits放大10倍后，概率分布变得**非常尖锐**（接近one-hot编码），类别0的概率接近100%，其他类别概率接近0。这说明**高温（实际对应低温度参数）会使Softmax更"硬"**，倾向于选择最大值。

**缩小10倍（z'' = [0.2, 0.1, 0.01]）：**

| 类别 | 原始Logit | 缩小10倍 z'' | $e^{z''}$ | Softmax概率 |
|:---:|:---:|:---:|:---:|:---:|
| 0 | 2.0 | 0.2 | **1.221** | **0.417** |
| 1 | 1.0 | 0.1 | **1.105** | **0.377** |
| 2 | 0.1 | 0.01 | **1.010** | **0.345** |

**步骤4答案：**

当logits缩小10倍后，概率分布变得**更加均匀**，三个类别的概率差异减小（约41.7%、37.7%、34.5%）。这说明**低温（实际对应高温度参数）会使Softmax更"软"**，概率分布更平缓。

In [3]:
# ==========================================
# 练习2：答案验证代码
# ==========================================

def softmax(z):
    """Softmax函数"""
    exp_z = np.exp(z - np.max(z))
    return exp_z / np.sum(exp_z)

print('=== 练习2：答案验证 ===')
print()

# 任务1：基础计算
z1 = np.array([2.0, 1.0, 0.1])
print('任务1：基础Softmax计算')
print(f'Logits: {z1}')
print(f'Exp values: {np.exp(z1)}')
print(f'Sum of exp: {np.sum(np.exp(z1)):.3f}')
print(f'Softmax: {softmax(z1)}')
print(f'预测类别: {np.argmax(softmax(z1))} (概率最高)')
print()

# 任务2：放大10倍
z2 = z1 * 10
print('任务2：放大10倍（高温/硬分布）')
print(f'Logits: {z2}')
print(f'Exp values: [{np.exp(20):.2e}, {np.exp(10):.2e}, {np.exp(1):.3f}]')
probs2 = softmax(z2)
print(f'Softmax: [{probs2[0]:.6f}, {probs2[1]:.6e}, {probs2[2]:.6e}]')
print()

# 任务3：缩小10倍
z3 = z1 / 10
print('任务3：缩小10倍（低温/软分布）')
print(f'Logits: {z3}')
print(f'Exp values: {np.exp(z3)}')
print(f'Sum of exp: {np.sum(np.exp(z3)):.3f}')
print(f'Softmax: {softmax(z3)}')
print()

print('✅ 验证完成！')

=== 练习2：答案验证 ===

任务1：基础Softmax计算
Logits: [2.  1.  0.1]
Exp values: [7.3890561  2.71828183 1.10517092]
Sum of exp: 11.213
Softmax: [0.65900114 0.24243297 0.09856589]
预测类别: 0 (概率最高)

任务2：放大10倍（高温/硬分布）
Logits: [20. 10.  1.]
Exp values: [4.85e+08, 2.20e+04, 2.718]
Softmax: [0.999955, 4.539787e-05, 5.602542e-09]

任务3：缩小10倍（低温/软分布）
Logits: [0.2  0.1  0.01]
Exp values: [1.22140276 1.10517092 1.01005017]
Sum of exp: 3.337
Softmax: [0.36605947 0.33122431 0.30271622]

✅ 验证完成！


### ✅ 思考题答案

**问题1：Softmax为什么叫Softmax？与Hardmax有什么区别？**

**答案：**

- **Hardmax**：直接将最大值位置设为1，其余为0。例如输入[2, 1, 0] → 输出[1, 0, 0]。
  - 优点：简单直观
  - 缺点：不可微分，无法使用梯度下降优化

- **Softmax**：通过指数函数将输入转换为概率分布。例如输入[2, 1, 0] → 输出[0.665, 0.244, 0.091]。
  - "Soft"体现在：所有类别都有非零概率，不是绝对化的选择
  - 优点：平滑可导，可以用链式法则计算梯度，适合神经网络训练
  - "Max"体现在：通过指数放大差距，最高得分获得最高概率

---

**问题2：在神经网络训练中，为什么温度（temperature）参数很重要？**

**答案：**

温度参数τ控制Softmax的"软硬"程度：

- **τ → 0**（低温）：Softmax趋近Hardmax，输出接近one-hot，模型非常"自信"
- **τ = 1**（标准）：正常的Softmax行为
- **τ → ∞**（高温）：Softmax趋近均匀分布，模型非常"犹豫"

温度参数的定义来源于玻尔兹曼分布：
$$P_{i}=\frac{e^{-E_{i} /(k T)}}{\sum_{j} e^{-E_{j} /(k T)}}$$

**实际应用：**
- 训练初期可以使用较高温度（τ > 1），让模型充分探索各个类别
- 训练后期降低温度（τ < 1），让模型更确定自己的选择
- 上述即是，模拟退火的训练方式。
- 知识蒸馏中常用温度参数来平滑教师模型的输出，提供更多训练信号

**温度系数的选择范围：**
- 常用默认值：0.7 ~ 1.0，在创造性和稳定性之间取得较好的平衡。输出既不会过于呆板重复，也不至于变成胡言乱语，适合日常对话、摘要、代码生成等大多数通用场景。
- 保守模式：0.1 ~ 0.3，模型几乎总是选择概率最高的词。输出非常稳定、刻板，重复性高，适合需要精确回答的任务。
- 创意模式：1.0 ~ 1.2，概率分布变得平滑，模型开始愿意选择那些原本概率较低的词。输出开始变得大胆、有想象力，但逻辑性和事实性会有所下降。
- 实验模式：1.3 ~ 1.5，输出的随机性显著增强，可能会出现逻辑断裂、语法混乱或严重的幻觉。

**温度系数通常可与Top p系数组合搭配使用：**
- 温度和 Top-p 的搭配，本质是在 “确定性 vs. 多样性” 之间找平衡。
- Top-p（核采样）：选出一组“概率和”刚好达到 p 的“最小”词集合。选择动态数量的词，直到累积概率达到p。Top-p 通常比 Top-k 更常用，因为它能自适应模型当前时刻的置信度，逻辑上更合理。
- 先调温度：快速控制“胆量”。发现重复、僵化就升温；发现逻辑乱、幻觉多就降温。
- 再动Top-p：微调“词汇池”。高温下输出太怪异，就适当降低 Top-p（比如从 0.95 降到 0.8），剔除极端词；输出太无聊，就升高 Top-p，让词汇选择更丰富。
- p 较小（如 0.3 - 0.5）：相当于“高精度筛选”。只保留概率最高的极少部分。输出非常集中、安全，适合逻辑严谨的任务。
- p 中等（如 0.7 - 0.9）：在安全和多样之间平衡。既过滤掉了概率极低的“长尾噪声词”，又保留了有一定可能性的合理选项。
- p 较大（如 0.95 - 1.0）：几乎接近全量词汇表。p=1.0 等同于不截断，但通常也会过滤掉那些概率为 0.0001% 的极端生僻词。


---

## <font color="#FFEA00" >答案速查表</font>

### 练习1：Logit与Sigmoid转换

| p | z | σ(z) |
|:---:|:---:|:---:|
| 0.5 | 0 | 0.5 |
| 0.8 | 1.386 | 0.8 |
| 0.2 | -1.386 | 0.2 |
| 0.5 | 0 | 0.5 |
| 0.881 | 2 | 0.881 |
| 0.269 | -1 | 0.269 |

### 练习2：Softmax计算

**原始logits [2.0, 1.0, 0.1]：**
- 概率：[0.659, 0.242, 0.099]
- 预测类别：0

**放大10倍 [20, 10, 1]：**
- 概率：[≈1.0, ≈0, ≈0]（尖锐分布）

**缩小10倍 [0.2, 0.1, 0.01]：**
- 概率：[0.417, 0.377, 0.345]（平缓分布）